chess-AI/experimental-work/mcts.py

```diff
@@ -1,73 +1,161 @@
+        self.exploration = 1 # sets exploration rate
+        self.p_values = {} # stores the probabilities computed by NN
 
-        if s not in visited # if a state has not been visited then you must find the predictions made by the model
-            visited.add(s) # will mark as visisted
-            P[s],v = nnet.predict(s) # this is like the rollout phase but using the nueral network
-            return -value
-        
-        max_u, best_a = -float("inf"), -1 
+    def find_best_move(self,legal_moves,fen_string):
+
+        best_u = -float("inf")
+        best_move = -1
+
+        # Iterate through all legal moves at the state
+
+        for move in legal_moves:
+
+            # Get move in uci form
+            uci_move = move.uci()
+
+            # Check to see if the state has been added to Q
+            if fen_string in self.q_values:
+                if uci_move in self.q_values[fen_string]:
+                    q_value = self.q_values[fen_string][uci_move]
+                else:
+                    self.q_values[fen_string][uci_move] = 0
+                    q_value = 0
+            else:
+                self.q_values[fen_string]={}
+                self.q_values[fen_string][uci_move] = 0
+                q_value = 0
+
+
+            # Check to see if state has been added to N
+            if fen_string in self.n_values:
+                if uci_move in self.n_values[fen_string]:
+                    n_value = self.n_values[fen_string][uci_move]
+                else:
+                    self.n_values[fen_string][uci_move] = 1
+                    n_value = float("inf")
+            else:
+                self.n_values[fen_string]={}
+                self.n_values[fen_string][uci_move] = 1
+                n_value = float("inf")
+
+            """
+            NOTE: still unsure about the n values
+            not sure if I should set the n values to infinite if a node
+            has not been visited
+            Also do not know what N should be initial
+            """
+
+            # Calculate U based on the UCB formula
+            u_value = q_value + (self.exploration * self.p_values[fen_string][uci_move])
+            u_value = u_value * sqrt(sum(self.n_values[fen_string].values())/n_value)
+
+            """
+            NOTE: Look at different formulas for calculating U
+            There could be better formula for evaluating an action
+            """
+
+            # Want to choose the move with the highest u value
+            if u_value > best_u:
+                best_u = u_value
+                best_move = move # move with best u value
+
+        return best_move
 
-        for a in game.getvalidActions(s): # finds the valid actions from current state
 
-            u = Q[s][a] + c_puct*P[s][a]*sqrt(sum(N[s]))/(1+N[s][a]) # formula for calculating the action to take
+    # Function to calculate Q and N based on the value and the move
+    def update_qn(self,fen_string,move,value):
 
```

Comment:
Should have a docstring here. Can move the comment above the function into the docstring.